### DISCUSSION TITLE
Week 5: July 15-19

### DISCUSSION THREAD

**Discussion by eeholmes** [SAFS-Varanasi-Internship/Summer-2024#19] [General]
Monday: Eli

* Working on the CHL data. 
* [Daily Level 3 chl data](https://data.marine.copernicus.eu/product/OCEANCOLOUR_GLO_BGC_L3_MY_009_103/files?subdataset=cmems_obs-oc_glo_bgc-plankton_my_l3-multi-4km_P1D_202311) id: cmems_obs-oc_glo_bgc-plankton_my_l3-multi-4km_P1D
[Daily Level 4 chl data](https://data.marine.copernicus.eu/product/OCEANCOLOUR_GLO_BGC_L4_MY_009_104/files?subdataset=cmems_obs-oc_glo_bgc-plankton_my_l4-gapfree-multi-4km_P1D_202311) id: cmems_obs-oc_glo_bgc-plankton_my_l4-gapfree-multi-4km_P1D

**Comment by eeholmes** [comment]
Compare these two extents:

* big lat=slice(32, -11.75), lon=slice(42,101.75)
* small lat=slice(35, -5.25), lon=slice(45,79.75)
*

**Comment by eeholmes** [comment]
@yifeihang Example creating a zarr dir. Shows how to append data to a zarr directory.

```
import xarray as xr
import pandas as pd
ds0 = xr.Dataset({'temperature': (['time'],  [50, 51, 52])}, coords={'time': pd.date_range('2000-01-01', periods=3)})
ds1 = xr.Dataset({'pressure': (['time'],  [50, 51, 52])}, coords={'time': pd.date_range('2000-01-01', periods=3)})
store = zarr.DirectoryStore('test.zarr')
ds0.to_zarr(store, mode='w')
ds1.to_zarr(store, mode='a')
xr.open_zarr(store)
```

**Comment by ShridharS19** [comment]
![orig1](https://github.com/user-attachments/assets/06a3e9d0-c41c-4705-8037-f01cdd03d3b9)
![orig2](https://github.com/user-attachments/assets/39bfc9fc-a778-4c7c-a19b-764dbc57aed2)
Improved PINN prediction
working on fine tuning and hyperparameter search

**Comment by eeholmes** [comment]
Wednesday. Eli

* got the chlorophyll gapfree and level 3 (gaps) products saved as a merged zarr file
* next step move to other JHub

<img width="764" alt="image" src="https://github.com/user-attachments/assets/0bee19b5-74c8-4c9b-a2a3-35955a8aa396">

**Reply by eeholmes** [reply]
<img width="687" alt="image" src="https://github.com/user-attachments/assets/8d82b466-df3b-4ac2-ae45-ef68d1e9abe9">

**Comment by eeholmes** [comment]
@yifeihang @ShridharS19 New data set with the new CHL layers `shared/data/IO.zarr`. `CHL_cmes-gapfree` is the `CHL` variable you were using (but wo the missing data in summer) and `CHL_cmes-level3` is the one with clouds. `CHL_cmes_cloud` is a flag with land, cloud and water for each image (in both CHL layers).

* `CHL_cmes-gapfree`: Gap-filled chlorophyll-a concentration (mg/m**3), from Oct 1997 (CMES_CHL)
* `CHL_cmes_uncertainty-gapfree`: Chlorophyll-a concentration uncertainty (%)  (CMES_CHL)
* `CHL_cmes_flags-gapfree`: 0=land, 1=observed, 2=interpolated, 3=NA (CMES_CHL)
* `CHL_cmes-level3`: Level 3 (means has gaps) chlorophyll-a concentration (mg/m**3). CHL_interp_my is created from this product. from Oct 1997 (CMES_CHL)
* `CHL_cmes_uncertainty-level3`: chlorophyll-a concentration uncertainty (%)  (CMES_CHL)
* `CHL_cmes_land`: 2=land, 0=water  (CMES_CHL)
* `CHL_cmes_cloud`: 2=land, 1=cloud, 0=water  (CMES_CHL)

**Reply by eeholmes** [reply]
![chl](https://github.com/user-attachments/assets/9f4c3b18-7928-43a8-8470-952db6944a5e)

**Reply by eeholmes** [reply]
```
# set up axes and color bar
#https://gradsaddict.blogspot.com/2019/11/python-tutorial-intro-to-cartopy.html
import xarray as xr
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import matplotlib.animation as animation
import matplotlib as mpl
import cartopy.crs as ccrs
from IPython.display import HTML

# load data
ds = xr.open_zarr("IO.zarr")
ds3=ds.sel(time=slice("2022-05-01", "2022-09-30"))

# set up axes
levels = range(10)
fig, axes = plt.subplots(figsize=[11,8], subplot_kw={'projection': ccrs.PlateCarree()})
axes.add_feature(cfeature.LAND,zorder=4,color='gray')
gl = axes.gridlines(draw_labels=True)
gl.xlines = False
gl.ylines = False
gl.top_labels = False
gl.right_labels = False
cmap = mpl.colormaps.get_cmap('viridis')  # viridis is the default colormap for imshow
cmap.set_bad(color='white')

# animate
def animate(time):
    ds3.isel(time=time)['CHL_cmes-level3'].plot(levels=levels, ax=axes, add_colorbar= False, cmap=cmap)

ani = animation.FuncAnimation(fig, animate, 10, interval=400, blit=False)

# add colorbar
mld=ds3.isel(time=0)['CHL_cmes-level3'].plot(levels=levels, ax=axes, add_colorbar= False, cmap=cmap);
cbar=fig.colorbar(mld)
cbar.set_label('CHL')

ani.save('chl.gif', writer='imagemagick', fps = 2) #Save animation as gif-file

HTML(ani.to_jshtml()) #Show the animation in the kernel
```

**Reply by eeholmes** [reply]
@yifeihang 

Creating the flag for pixels that are treated as cloud but that really have an observation so we can test.

Using 10 (or whatever) days ahead as the fake cloud means that clouds are realistic clouds from the right season.
```
import xarray as xr
ds = xr.open_zarr("/home/jovyan/shared/data/IO.zarr")
land = ds['CHL_cmes-cloud'].isel(time=0) == 2
cloud1 = ds['CHL_cmes-cloud'].isel(time=0) == 1
obs1 = ~cloud1 & ~land
# we use another day 10 days in future to provide a real 'fake' cloud
cloud2 = ds['CHL_cmes-cloud'].isel(time=10) == 1
obs1wfakecloud = obs1 & cloud2
```

**Comment by eeholmes** [comment]
Links to 'books'

* https://safs-varanasi-internship.github.io/indian-ocean-zarr/
* https://safs-varanasi-internship.github.io/2024-tutorials/

**Comment by eeholmes** [comment]
2nd GPU is not spawnin

<img width="581" alt="image" src="https://github.com/user-attachments/assets/13811b26-dda0-4e06-abf7-506ea28a2fe1">

pvc-1ee6c292-dfeb-4017-83b1-daedd6288a7e

**Reply by eeholmes** [reply]
@ShridharS19 Can you try again? I was able to log in to the GPU ok and didn't see any obvious issues with the GPU pods.
